### PR TITLE
feat: create colors hooks

### PR DESCRIPTION
**Pull request by leko91**
~~For now, we only have `GET`s, but these will be updated when the backend side is done. 
I would merge this without all the methods so we're not blocked and update the hooks once we have all the endpoints.~~
~~Only add color left now.~~
~~Missing create and delete palette.~~
Should be complete now! 🥳

### CODE REVIEW

File: src/react/useColorPalettes.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+export const useColorPalettes = (appBridge: AppBridgeBlock, colorPaletteIds?: number[]): UseColorPalettesReturnType => {
+    const [colorPalettes, setColorPalettes] = useState<FrontifyColorPalette[]>([]);
+
+    useEffect(() => {
+        const getColorPalettes = async () => {
+            try {
+                const colorPalettes: FrontifyColorPalette[] = await appBridge.getColorPalettes();
+
+                if (colorPaletteIds) {
+                    const filteredColorPalettes = colorPalettes.filter((colorPalette: FrontifyColorPalette) =>
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:20]: Prefer a `getColorPaletteById`, if it doesn't exist you can create it

File: src/react/useColors.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import { useEffect, useState } from 'react';
+import { AppBridgeBlock } from '../AppBridgeBlock';
+import { FrontifyColor } from '../types/FrontifyColor';
+
+export type UseColorsReturnType = {
+    colorsByPaletteId: FrontifyColor[];
+};
+
+export const useColors = (appBridge: AppBridgeBlock): UseColorsReturnType => {
```
Comment (SamuelAlev) [inline src/react/useColors.ts:11]:
```suggestion
export const useColors = (appBridge: AppBridgeBlock, colorPaletteId: number): UseColorsReturnType => {
```

Comment (SamuelAlev) [reply src/react/useColors.ts:11]: `useColors` should return the colors of a palette

File: src/react/useColors.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                const colors = await appBridge.getColorsByColorPaletteId(colorPaletteId);
+
+                setColorsByPaletteId(colors);
+            } catch (error) {
+                console.error('Error: ', error);
+            }
+        };
+
+        getColorsByColorPaletteId(appBridge.getBlockId());
+    }, []);
```
Comment (SamuelAlev) [inline src/react/useColors.ts:26]: Then you can watch for changes 
```suggestion
    }, [colorPaletteId]);
```

File: src/react/useColorPalettes.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                        colorPaletteIds.includes(colorPalette.id),
+                    );
+
+                    await Promise.all(
+                        filteredColorPalettes.map(
+                            async (colorPalette: FrontifyColorPalette) =>
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:26]:
```suggestion
                            async (colorPalette) =>
```
Type should be implicitely set already

**Comment by SamuelAlev** [comment]
tests? 😄

File: src/react/useColorPalettes.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        const getColorPalettes = async () => {
+            try {
+                const colorPalettes: FrontifyColorPalette[] = await appBridge.getColorPalettes();
+
+                if (colorPaletteIds) {
+                    const filteredColorPalettes = colorPalettes.filter((colorPalette: FrontifyColorPalette) =>
+                        colorPaletteIds.includes(colorPalette.id),
+                    );
+
+                    await Promise.all(
```
Comment (GoranRibic) [inline src/react/useColorPalettes.ts:24]: Why do you get colors only in the if statement?

File: src/react/useColorPalettes.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                } else {
+                    setColorPalettes(colorPalettes);
+                }
+            } catch (error) {
+                console.error('Error: ', error);
+            }
+        };
+
+        getColorPalettes();
+    }, []);
```
Comment (GoranRibic) [inline src/react/useColorPalettes.ts:41]: Can you please add colorPaletteIds and other dependencies in the dependency array.

File: src/react/useColorPalettes.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                    );
+
+                    await Promise.all(
+                        filteredColorPalettes.map(
+                            async (colorPalette: FrontifyColorPalette) =>
+                                (colorPalette.colors = await appBridge.getColorsByColorPaletteId(colorPalette.id)),
+                        ),
+                    );
+
+                    setColorPalettes(filteredColorPalettes);
```
Comment (GoranRibic) [inline src/react/useColorPalettes.ts:31]: Rewrite code such that you will only use setColorPalettes once. Two times is redundant, and it will look more readable if used once.

File: src/react/useColors.ts (hunk 3 above)
Comment (GoranRibic) [inline src/react/useColors.ts:26]: Please add missing dependencies.

File: src/react/useColors.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                const colors = await appBridge.getColorsByColorPaletteId(colorPaletteId);
+
+                setColorsByPaletteId(colors);
+            } catch (error) {
+                console.error('Error: ', error);
+            }
+        };
+
+        getColorsByColorPaletteId(colorPaletteId);
+    }, [appBridge]);
```
Comment (GoranRibic) [inline src/react/useColors.ts:26]: You are missing colorPaletteId as dependency.

**Review by GoranRibic — APPROVED** [review]

File: src/react/useColors.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        await appBridge.createColor(colorCreate);
+    };
+
+    const updateColor = async (colorId: number, colorPatch: FrontifyColorPatch) => {
+        await appBridge.updateColor(colorId, colorPatch);
+    };
+
+    const deleteColor = async (colorId: number) => {
+        await appBridge.deleteColor(colorId);
+    };
```
Comment (SamuelAlev) [inline src/react/useColors.ts:41]: Calling the methods don't update `colorsByPaletteId`

File: src/react/useColorPalettes.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+                console.error('Error: ', error);
+            }
+        };
+
+        getColorPalettes();
+    }, [appBridge, colorPaletteIds]);
+
+    const updateColorPalette = async (colorPaletteId: number, colorPalettePatch: FrontifyColorPalettePatch) => {
+        await appBridge.updateColorPalette(colorPaletteId, colorPalettePatch);
+    };
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:31]: Calling the method don't update `colorPalettes`

File: src/react/useColors.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+        setColorsByPaletteId((previousState) =>
+            previousState.map((color, index) => (index === modifiedColorIndex ? updatedColor : color)),
+        );
+    };
+
+    const deleteColor = async (colorId: number) => {
+        await appBridge.deleteColor(colorId);
+
+        setColorsByPaletteId((previousState) => previousState.filter((color) => color.id !== colorId));
```
Comment (SamuelAlev) [inline src/react/useColors.ts:51]: This doesn't work if you have 2 times the hooks in different blocks, you can check `useBlockSettings` or `useBlockAssets` to see how we do it (it uses custom events)

File: src/react/useColorPalettes.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        const updatedColorPalette = await appBridge.updateColorPalette(colorPaletteId, colorPalettePatch);
+
+        const modifiedColorPalette = colorPalettes.find((colorPalette) => colorPalette.id === colorPaletteId);
+        const modifiedColorPaletteIndex = modifiedColorPalette && colorPalettes.indexOf(modifiedColorPalette);
+
+        setColorPalettes((previousState) =>
+            previousState.map((colorPalette, index) =>
+                index === modifiedColorPaletteIndex ? updatedColorPalette : colorPalette,
+            ),
+        );
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:39]: This doesn't work if you have 2 times the hooks in different blocks, you can check `useBlockSettings` or `useBlockAssets` to see how we do it (it uses custom events)

File: src/react/useColorPalettes.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    };
+
+    return {
+        colorPalettes,
+        updateColorPalette,
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:44]: We miss the create and delete color palette

File: src/react/useColors.ts (hunk 11 above)
Comment (leko91) [reply src/react/useColors.ts:51]: Hm, locally it has no issues for me, but I'll change it to use the custom events

File: src/react/useColorPalettes.ts (hunk 13 above)
Comment (leko91) [reply src/react/useColorPalettes.ts:44]: There are no endpoints for that, that happens on block creation/deletion, right?

File: src/react/useColors.ts (hunk 11 above)
Comment (SamuelAlev) [reply src/react/useColors.ts:51]: Let me rephrase, if you have 2 blocks and use the `create` in one of them, the other block won't be notified and you would need to refresh the page to see the change in the other block

File: src/react/useColorPalettes.ts (hunk 13 above)
Comment (SamuelAlev) [reply src/react/useColorPalettes.ts:44]: Not for custom blocks

File: src/react/useColorPalettes.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        emitUpdatedColorPalettes();
+    };
+
+    const deleteColorPalette = async (colorPaletteId: number) => {
+        await appBridge.deleteColorPalette(colorPaletteId);
+
+        emitUpdatedColorPalettes();
+    };
+
+    const getDownloadColorKitLink = (selectedColorPalettes: number[]) => {
```
Comment (SamuelAlev) [inline src/react/useColorPalettes.ts:82]:
```suggestion
    const downloadColorKit = (selectedColorPalettes: number[]) => {
```
Same here 😅

**Review by SamuelAlev — APPROVED** [review]
